README.md
# Swedish Personal Identity Number Library

![CI](https://github.com/Patberg/swedish-personal-identity-number/actions/workflows/main.yml/badge.svg) ![CD](https://github.com/Patberg/swedish-personal-identity-number/actions/workflows/publish.yml/badge.svg) ![npm](https://img.shields.io/npm/v/swedish-personal-identity-number?logo=npm)

## Overview

This library, named `swedish-personal-identity-number`, offers a robust solution for managing Swedish personal identity numbers. It ensures that input data matches the expected format and offers detailed insights about the parsed number, including date of birth, gender, and place of birth.

For a deeper understanding of Swedish personal identity numbers, refer to the [Wikipedia documentation](<https://en.wikipedia.org/wiki/Personal_identity_number_(Sweden)>).

## Installation

To use this library, you can install it using npm, yarn, or pnpm:

### npm

```bash
npm install swedish-personal-identity-number
```

### Yarn

```bash
yarn add swedish-personal-identity-number
```

### pnpm

```bash
pnpm add swedish-personal-identity-number
```

## Module Support

This library supports both **ES Modules** and **CommonJS**.

Then, depending on your environment and setup, you can import it:

```javascript
// Using ES Modules
import { PersonalIdentityNumber } from "swedish-personal-identity-number";

// Using CommonJS
const { PersonalIdentityNumber } = require("swedish-personal-identity-number");
```

## Key Features

1. **Format Validation** - Quickly validate the format of a given personal identity number.
2. **Checksum(Luhn) Validation** - Ensures that the checksum of the number is valid.
3. **Gender Identification** - Determines gender based on the personal identity number.
4. **Date and Place of Birth** - Parses the date of birth and potentially the place of birth if it's available.
5. **Coordination Number Identification** - Determines if a number is a coordination number rather than a standard personal identity number.

## Supported Formats

The `parse` and `tryParse` methods of the `PersonalIdentityNumber` class can handle a variety of formats commonly used for Swedish personal identity numbers. Specifically, the library supports the following formats:

1. Full with century and without hyphen: `YYYYMMDDXXXX`  
   Example: `199911232388`

2. Full with century and with hyphen: `YYYYMMDD-XXXX`  
   Example: `19991123-2388`

3. Short without century and without hyphen: `YYMMDDXXXX`  
   Example: `9911232388`

4. Short without century and with hyphen: `YYMMDD-XXXX`  
   Example: `991123-2388`

5. Short without century and with plus(indicating that the person was born in the previous century): `YYMMDD+XXXX`  
   Example: `121212+1212`

---

## API Documentation

### Class: PersonalIdentityNumber

#### Static Methods

##### `parse(personalIdentityNumber: string): PersonalIdentityNumber`

- Parses the given Swedish personal identity number string.
- Throws an error if the input number is invalid or doesn't pass the Luhn check.
- Returns an instance of `PersonalIdentityNumber`.

##### `tryParse(personalIdentityNumber: string): PersonalIdentityNumber | null`

- Tries to parse the given string.
- Returns an instance of `PersonalIdentityNumber` if successful, otherwise returns null.

#### Instance Methods

##### `toString(): string`

- Converts the current instance of `PersonalIdentityNumber` to a string representation with century and with hyphen: `YYYYMMDD-XXXX`  
  Example: `19991123-2388`

##### `toShortString(): string`

- Returns a shortened string representation of the personal identity number by slicing off the century part.
  Example: `991123-2388`

##### `equals(other: PersonalIdentityNumber): boolean`

- Compares the current instance with another `PersonalIdentityNumber` instance.
- Returns `true` if both instances represent the same number, otherwise `false`.

#### Properties

- `dateOfBirth: Date` - Represents the date of birth extracted from the personal identity number.
- `digits: number` - The last four digits of the personal identity number.
- `gender: "Male" | "Female"` - Gender derived from the personal identity number.
- `placeOfBirth: PlaceOfBirth` - Represents the place of birth which might be unknown for numbers after 1990.
- `isCoordinationNumber: boolean` - Indicates if the number is a coordination number.

The supported places of birth include:

    Stockholms stad
    Stockholms län
    Uppsala län
    Södermanlands län
    Östergötlands län
    Jönköpings län
    Kronobergs län
    Kalmar län
    Gotlands län
    Blekinge län
    Kristianstads län
    Malmöhus län
    Hallands län
    Göteborgs och Bohus län
    Älvsborgs län
    Skaraborgs län
    Värmlands län
    Örebro län
    Västmanlands län
    Kopparbergs län
    Gävleborgs län
    Västernorrlands län
    Jämtlands län
    Västerbottens län
    Norrbottens län

Additionally, there is an "Unknown" option that is used when the place of birth is not determined.

## Usage Example

```javascript
import { PersonalIdentityNumber } from "path-to-library";

const pin = "YYMMDD-XXXX";

const result = PersonalIdentityNumber.tryParse(pin);

if (result) {
  console.log("Date of Birth:", result.dateOfBirth);
  console.log("Gender:", result.gender);
  console.log("Place of Birth:", result.placeOfBirth);
  console.log("Is Coordination Number:", result.isCoordinationNumber);
} else {
  console.error("Invalid Personal Identity Number.");
}
```

## Repository & Author

- **GitHub Repository**: [swedish-personal-identity-number](https://github.com/Patberg/swedish-personal-identity-number.git)
- **Author**: [Patrik Bergström](https://github.com/Patberg)
- **License**: MIT

## Contribution

For any issues, improvements, or feature requests, please create an issue or pull request on the [GitHub repository](https://github.com/Patberg/swedish-personal-identity-number.git). Feedback is highly appreciated!

---


CHANGELOG.md
# swedish-personal-identity-number

## 1.0.0

### Major Changes

- 7fe3e76: Initial version


src/placeOfBirth.test.ts
import { describe, expect, it } from "vitest";
import { getPlaceOfBirth } from "./placeOfBirth";

describe("getPlaceOfBirth", () => {
  it("should return 'Stockholms stad' for digits 00", () => {
    expect(getPlaceOfBirth("00")).toBe("Stockholms stad");
  });

  it("should return 'Stockholms län' for digits 10-13", () => {
    expect(getPlaceOfBirth("10")).toBe("Stockholms län");
    expect(getPlaceOfBirth("11")).toBe("Stockholms län");
    expect(getPlaceOfBirth("12")).toBe("Stockholms län");
    expect(getPlaceOfBirth("13")).toBe("Stockholms län");
  });

  it("should return 'Uppsala län' for digits 14-15", () => {
    expect(getPlaceOfBirth("14")).toBe("Uppsala län");
    expect(getPlaceOfBirth("15")).toBe("Uppsala län");
  });

  it("should return 'Södermanlands län' for digits 16-18", () => {
    expect(getPlaceOfBirth("16")).toBe("Södermanlands län");
    expect(getPlaceOfBirth("17")).toBe("Södermanlands län");
    expect(getPlaceOfBirth("18")).toBe("Södermanlands län");
  });

  it("should return 'Östergötlands län' for digits 19-23", () => {
    expect(getPlaceOfBirth("19")).toBe("Östergötlands län");
    expect(getPlaceOfBirth("20")).toBe("Östergötlands län");
    expect(getPlaceOfBirth("21")).toBe("Östergötlands län");
    expect(getPlaceOfBirth("22")).toBe("Östergötlands län");
    expect(getPlaceOfBirth("23")).toBe("Östergötlands län");
  });

  it("should return 'Jönköpings län' for digits 24-26", () => {
    expect(getPlaceOfBirth("24")).toBe("Jönköpings län");
    expect(getPlaceOfBirth("25")).toBe("Jönköpings län");
    expect(getPlaceOfBirth("26")).toBe("Jönköpings län");
  });

  it("should return 'Kronobergs län' for digits 27-28", () => {
    expect(getPlaceOfBirth("27")).toBe("Kronobergs län");
    expect(getPlaceOfBirth("28")).toBe("Kronobergs län");
  });

  it("should return 'Kalmar län' for digits 29-31", () => {
    expect(getPlaceOfBirth("29")).toBe("Kalmar län");
    expect(getPlaceOfBirth("30")).toBe("Kalmar län");
    expect(getPlaceOfBirth("31")).toBe("Kalmar län");
  });

  it("should return 'Gotlands län' for digit 32", () => {
    expect(getPlaceOfBirth("32")).toBe("Gotlands län");
  });

  it("should return 'Blekinge län' for digits 33-34", () => {
    expect(getPlaceOfBirth("33")).toBe("Blekinge län");
    expect(getPlaceOfBirth("34")).toBe("Blekinge län");
  });

  it("should return 'Kristianstads län' for digits 35-38", () => {
    expect(getPlaceOfBirth("35")).toBe("Kristianstads län");
    expect(getPlaceOfBirth("36")).toBe("Kristianstads län");
    expect(getPlaceOfBirth("37")).toBe("Kristianstads län");
    expect(getPlaceOfBirth("38")).toBe("Kristianstads län");
  });

  it("should return 'Malmöhus län' for digits 39-45", () => {
    expect(getPlaceOfBirth("39")).toBe("Malmöhus län");
    expect(getPlaceOfBirth("40")).toBe("Malmöhus län");
    expect(getPlaceOfBirth("41")).toBe("Malmöhus län");
    expect(getPlaceOfBirth("42")).toBe("Malmöhus län");
    expect(getPlaceOfBirth("43")).toBe("Malmöhus län");
    expect(getPlaceOfBirth("44")).toBe("Malmöhus län");
    expect(getPlaceOfBirth("45")).toBe("Malmöhus län");
  });

  it("should return 'Hallands län' for digits 46-47", () => {
    expect(getPlaceOfBirth("46")).toBe("Hallands län");
    expect(getPlaceOfBirth("47")).toBe("Hallands län");
  });

  it("should return 'Göteborgs och Bohus län' for digits 48-54", () => {
    expect(getPlaceOfBirth("48")).toBe("Göteborgs och Bohus län");
    expect(getPlaceOfBirth("49")).toBe("Göteborgs och Bohus län");
    expect(getPlaceOfBirth("50")).toBe("Göteborgs och Bohus län");
    expect(getPlaceOfBirth("51")).toBe("Göteborgs och Bohus län");
    expect(getPlaceOfBirth("52")).toBe("Göteborgs och Bohus län");
    expect(getPlaceOfBirth("53")).toBe("Göteborgs och Bohus län");
    expect(getPlaceOfBirth("54")).toBe("Göteborgs och Bohus län");
  });

  it("should return 'Älvsborgs län' for digits 55-58", () => {
    expect(getPlaceOfBirth("55")).toBe("Älvsborgs län");
    expect(getPlaceOfBirth("56")).toBe("Älvsborgs län");
    expect(getPlaceOfBirth("57")).toBe("Älvsborgs län");
    expect(getPlaceOfBirth("58")).toBe("Älvsborgs län");
  });

  it("should return 'Skaraborgs län' for digits 59-61", () => {
    expect(getPlaceOfBirth("59")).toBe("Skaraborgs län");
    expect(getPlaceOfBirth("60")).toBe("Skaraborgs län");
    expect(getPlaceOfBirth("61")).toBe("Skaraborgs län");
  });

  it("should return 'Värmlands län' for digits 62-64", () => {
    expect(getPlaceOfBirth("62")).toBe("Värmlands län");
    expect(getPlaceOfBirth("63")).toBe("Värmlands län");
    expect(getPlaceOfBirth("64")).toBe("Värmlands län");
  });

  it("should return 'Örebro län' for digits 66-68", () => {
    expect(getPlaceOfBirth("66")).toBe("Örebro län");
    expect(getPlaceOfBirth("67")).toBe("Örebro län");
    expect(getPlaceOfBirth("68")).toBe("Örebro län");
  });

  it("should return 'Västmanlands län' for digits 69-70", () => {
    expect(getPlaceOfBirth("69")).toBe("Västmanlands län");
    expect(getPlaceOfBirth("70")).toBe("Västmanlands län");
  });

  it("should return 'Kopparbergs län' for digits 71-73", () => {
    expect(getPlaceOfBirth("71")).toBe("Kopparbergs län");
    expect(getPlaceOfBirth("72")).toBe("Kopparbergs län");
    expect(getPlaceOfBirth("73")).toBe("Kopparbergs län");
  });

  it("should return 'Gävleborgs län' for digits 75-77", () => {
    expect(getPlaceOfBirth("75")).toBe("Gävleborgs län");
    expect(getPlaceOfBirth("76")).toBe("Gävleborgs län");
    expect(getPlaceOfBirth("77")).toBe("Gävleborgs län");
  });

  it("should return 'Västernorrlands län' for digits 78-81", () => {
    expect(getPlaceOfBirth("78")).toBe("Västernorrlands län");
    expect(getPlaceOfBirth("79")).toBe("Västernorrlands län");
    expect(getPlaceOfBirth("80")).toBe("Västernorrlands län");
    expect(getPlaceOfBirth("81")).toBe("Västernorrlands län");
  });

  it("should return 'Jämtlands län' for digits 82-84", () => {
    expect(getPlaceOfBirth("82")).toBe("Jämtlands län");
    expect(getPlaceOfBirth("83")).toBe("Jämtlands län");
    expect(getPlaceOfBirth("84")).toBe("Jämtlands län");
  });

  it("should return 'Västerbottens län' for digits 85-88", () => {
    expect(getPlaceOfBirth("85")).toBe("Västerbottens län");
    expect(getPlaceOfBirth("86")).toBe("Västerbottens län");
    expect(getPlaceOfBirth("87")).toBe("Västerbottens län");
    expect(getPlaceOfBirth("88")).toBe("Västerbottens län");
  });

  it("should return 'Norrbottens län' for digits 89-92", () => {
    expect(getPlaceOfBirth("89")).toBe("Norrbottens län");
    expect(getPlaceOfBirth("90")).toBe("Norrbottens län");
    expect(getPlaceOfBirth("91")).toBe("Norrbottens län");
    expect(getPlaceOfBirth("92")).toBe("Norrbottens län");
  });

  it("should return 'Unknown' for digits outside the valid range", () => {
    expect(getPlaceOfBirth("93")).toBe("Unknown");
    expect(getPlaceOfBirth("99")).toBe("Unknown");
  });

  it("should throw an error if digits is less than 2 characters long", () => {
    expect(() => getPlaceOfBirth("")).toThrow("Digits must be at least 2 characters long");
    expect(() => getPlaceOfBirth("1")).toThrow("Digits must be at least 2 characters long");
  });
});
// END: qm8d4b3jv5f7


src/personalIdentityNumber.test.ts
import { describe, expect, it } from "vitest";
import { PersonalIdentityNumber } from "./personalIdentityNumber";

describe("personalIdentityNumber", () => {
  it("undefined argument should throw error", () => {
    expect(() => PersonalIdentityNumber.parse(undefined!)).toThrow("personalIdentityNumber is undefined");
  });

  it("empty string argument should throw error", () => {
    expect(() => PersonalIdentityNumber.parse("")).toThrow("personalIdentityNumber is not valid");
  });

  it.each(["191212121212", "19121212-1212", "1212121212", "121212-1212"])(
    "valid personal number %s should not throw error",
    (personalNumber) => {
      expect(() => PersonalIdentityNumber.parse(personalNumber)).not.toThrow();
    },
  );

  it.each(["19920873-2389", "199208732389", "9208732389", "920873-2389"])(
    "valid co-ordination number %s should not throw error",
    (personalNumber) => {
      expect(() => PersonalIdentityNumber.parse(personalNumber)).not.toThrow();
    },
  );

  it.each(["199911232388", "19991123-2388", "9911232388", "19991123-2388", "991123 2388"])(
    "should be able to parse a valid personal number",
    (input) => {
      const personalNumber = PersonalIdentityNumber.parse(input);
      expect(personalNumber.dateOfBirth).toEqual(new Date(1999, 10, 23));
      expect(personalNumber.digits).toEqual(2388);
      expect(personalNumber.isCoordinationNumber).toEqual(false);
    },
  );

  it.each(["19920873-2389", "199208732389", "9208732389", "920873-2389"])(
    "should be able to parse a valid co-ordination number",
    (input) => {
      const personalNumber = PersonalIdentityNumber.parse(input);
      expect(personalNumber.dateOfBirth).toEqual(new Date(1992, 7, 13));
      expect(personalNumber.digits).toEqual(2389);
      expect(personalNumber.isCoordinationNumber).toEqual(true);
    },
  );

  it.each(["199911232388", "19991123-2388", "9911232388", "991123-2388"])(
    "should be able to convert valid personal numbers to a string uniform string format",
    (input) => {
      const personalNumber = PersonalIdentityNumber.parse(input);
      expect(personalNumber.toString()).toEqual("19991123-2388");
    },
  );

  it.each(["19920873-2389", "199208732389", "9208732389", "920873-2389"])(
    "should be able to convert valid temporary personal numbers to a string uniform string format",
    (input) => {
      const personalNumber = PersonalIdentityNumber.parse(input);
      expect(personalNumber.toString()).toEqual("19920873-2389");
    },
  );
  it.each(["", "foobar", "110873-1319"])("should return null for invalid personal identity numbers when using tryParse", (input) => {
    const personalNumber = PersonalIdentityNumber.tryParse(input);
    expect(personalNumber).toEqual(null);
  });

  it.each(["199911232388", "19991123-2388", "9911232388", "19991123-2388"])(
    "should be able to safely parse a valid personal number using tryparse",
    (input) => {
      const personalNumber = PersonalIdentityNumber.tryParse(input);
      expect(personalNumber!.dateOfBirth).toEqual(new Date(1999, 10, 23));
      expect(personalNumber!.digits).toEqual(2388);
      expect(personalNumber!.isCoordinationNumber).toEqual(false);
    },
  );

  it("should detect Male gender", () => {
    const personalNumber = PersonalIdentityNumber.parse("121212-1212");
    expect(personalNumber!.gender).toEqual("Male");
  });

  it("should detect Female gender", () => {
    const personalNumber = PersonalIdentityNumber.parse("19991123-2388");
    expect(personalNumber!.gender).toEqual("Female");
  });

  it("should be able to return a correct short string", () => {
    const personalNumber = PersonalIdentityNumber.parse("19991123-2388");
    expect(personalNumber!.toShortString()).toEqual("991123-2388");
  });

  it("should return unknown birth place if born in 1990 or later", () => {
    const personalNumber = PersonalIdentityNumber.parse("19991123-2388");
    expect(personalNumber!.placeOfBirth).toEqual("Unknown");
  });

  it("should return birth place if born before 1990", () => {
    const personalNumber = PersonalIdentityNumber.parse("198912242388");
    expect(personalNumber!.placeOfBirth).toEqual("Östergötlands län");
  });

  it("should return uknown birth place for a co-ordination even if the person was born before 1990", () => {
    const personalNumber = PersonalIdentityNumber.parse("197802662390");
    expect(personalNumber!.placeOfBirth).toEqual("Unknown");
  });

  it("equality should return true for instances with the same personal identity number", () => {
    const firstPersonalNumber = PersonalIdentityNumber.parse("198912242388");
    const secondPersonalNumber = PersonalIdentityNumber.parse("891224-2388");

    expect(firstPersonalNumber.equals(secondPersonalNumber)).toEqual(true);
  });

  it("equality should return false for instances with different personal identity numbers", () => {
    const firstPersonalNumber = PersonalIdentityNumber.parse("198912242388");
    const secondPersonalNumber = PersonalIdentityNumber.parse("19991123-2388");

    expect(firstPersonalNumber.equals(secondPersonalNumber)).toEqual(false);
  });

  it("should support '+' notation for persons older than 100 years", () => {
    const personalIdentityNumber = PersonalIdentityNumber.parse("121212+1212");
    expect(personalIdentityNumber.dateOfBirth.getFullYear()).toEqual(1912);
  });
});


src/personalIdentityNumber.ts
import { daysInMonth } from "./dateUtilities";
import { luhnCheck } from "./luhn";
import { PlaceOfBirth, getPlaceOfBirth } from "./placeOfBirth";

/**
 * Represents a Swedish personal identity number.
 */
export class PersonalIdentityNumber {
  static PERSONAL_IDENTITY_NUMBER_REGEX = /^(\d{2}){0,1}(\d{2})(\d{2})(\d{2})([+\-\s]?)(\d{3})(\d)/;

  /**
   * Date of birth of the person.
   */
  readonly dateOfBirth: Date;

  /**
   * Four-digit personal code.
   */
  readonly digits: number;

  /**
   * Gender of the person, can be either "Male" or "Female".
   */
  readonly gender: "Male" | "Female";

  /**
   * The region or municipality of birth.
   */
  readonly placeOfBirth: PlaceOfBirth;

  /**
   * Indicates if the personal identity number is a co-ordination number.
   */
  readonly isCoordinationNumber: boolean;

  private constructor(
    dateOfBirth: Date,
    digits: number,
    gender: "Male" | "Female",
    placeOfBirth: PlaceOfBirth = "Unknown",
    isCoordinationNumber: boolean = false,
  ) {
    this.dateOfBirth = dateOfBirth;
    this.digits = digits;
    this.gender = gender;
    this.placeOfBirth = placeOfBirth;
    this.isCoordinationNumber = isCoordinationNumber;
  }

  /**
   * Parses a string representation of a personal identity number and returns a new `PersonalIdentityNumber` instance.
   * Throws an error if the provided string is not a valid personal identity number.
   *
   * @param personalIdentityNumber - The string representation of a personal identity number. The string can be in any of the following formats: YYYYMMDD-XXXX, YYMMDD-XXXX, YYYYMMDDXXXX, YYMMDDXXXX, YYMMDD+XXXX, YYMMDD XXXX, YYMMDDXXXX.
   * @returns A new `PersonalIdentityNumber` instance.
   * @throws {Error} Throws an error if the argument is not a valid personal identity number.
   */
  static parse(personalIdentityNumber: string): PersonalIdentityNumber {
    if (personalIdentityNumber === undefined) {
      throw new Error("personalIdentityNumber is undefined");
    }

    const match = personalIdentityNumber.match(PersonalIdentityNumber.PERSONAL_IDENTITY_NUMBER_REGEX);

    if (match === null) {
      throw new Error("personalIdentityNumber is not valid");
    }

    if (match.length < 1 || match.length < 7) {
      throw new Error("personalIdentityNumber is not valid");
    }

    const decade = parseInt(match[2]);

    let century: string;

    if (match[1]) {
      century = match[1];
    } else {
      let born = new Date().getFullYear() - decade;
      if (match[5].length !== 0 && match[5] === "+") {
        born -= 100;
      }

      century = born.toString().substring(0, 2);
    }

    const month = match[3];
    const day = match[4];
    const digits = `${match[6]}${match[7]}`;

    if (!luhnCheck(`${decade}${month}${day}${digits}`)) {
      throw new Error(`The checksum ${digits} is not a valid checksum for the social seurity number ${personalIdentityNumber}`);
    }

    const parsedDay = parseInt(day);
    const parsedYear = parseInt(`${century}${decade}`);
    const parsedMonth = parseInt(month) - 1;

    const daysInBirthMonth = daysInMonth(parsedYear, parsedMonth);

    const gender = parseInt(digits[2]) % 2 === 0 ? "Female" : "Male";

    if (parsedDay <= daysInBirthMonth) {
      return new PersonalIdentityNumber(
        new Date(parsedYear, parsedMonth, parsedDay),
        parseInt(digits),
        gender,
        parsedYear < 1990 ? getPlaceOfBirth(digits) : "Unknown",
      );
    }

    const birthDay = parsedDay - 60;

    if (birthDay <= 0 || birthDay > daysInBirthMonth) {
      throw new Error(`The number '${personalIdentityNumber}' is not a valid social security number.`);
    }

    return new PersonalIdentityNumber(new Date(parsedYear, parsedMonth, birthDay), parseInt(digits), gender, "Unknown", true);
  }

  /**
   *
   * @param personalIdentityNumber - The string representation of a personal identity number. The string can be in any of the following formats: YYYYMMDD-XXXX, YYMMDD-XXXX, YYYYMMDDXXXX, YYMMDDXXXX, YYMMDD+XXXX, YYMMDD XXXX, YYMMDDXXXX.
   * @returns A new `PersonalIdentityNumber` instance or null if the argument is not a valid personal identity number.
   */
  static tryParse(personalIdentityNumber: string): PersonalIdentityNumber | null {
    try {
      return PersonalIdentityNumber.parse(personalIdentityNumber);
    } catch (error) {
      return null;
    }
  }

  /**
   * Returns a string representation of the personal identity number in the format YYYYMMDD-XXXX.
   *
   */

  toString(): string {
    if (this.isCoordinationNumber) {
      const formattedDate = `${this.dateOfBirth.getFullYear()}${(this.dateOfBirth.getMonth() + 1).toString().padStart(2, "0")}`;
      const day = this.dateOfBirth.getDate() + 60;
      const checkSum = this.digits.toString().padStart(4, "0");
      return `${formattedDate}${day}-${checkSum}`;
    } else {
      const formattedDate = `${this.dateOfBirth.getFullYear()}${(this.dateOfBirth.getMonth() + 1)
        .toString()
        .padStart(2, "0")}${this.dateOfBirth.getDate().toString().padStart(2, "0")}`;
      const checkSum = this.digits.toString().padStart(4, "0");
      return `${formattedDate}-${checkSum}`;
    }
  }

  /*
   * Returns a short string representation of the personal identity number in the format YYMMDD-XXXX.
   */

  toShortString = (): string => this.toString().slice(2);

  equals = (other: PersonalIdentityNumber): boolean => this.toString() === other.toString();
}


src/placeOfBirth.ts
/**
 * Represents the possible places of birth for a person in Sweden.
 */
export type PlaceOfBirth =
  | "Stockholms stad"
  | "Stockholms län"
  | "Uppsala län"
  | "Södermanlands län"
  | "Östergötlands län"
  | "Jönköpings län"
  | "Kronobergs län"
  | "Kalmar län"
  | "Gotlands län"
  | "Blekinge län"
  | "Kristianstads län"
  | "Malmöhus län"
  | "Hallands län"
  | "Göteborgs och Bohus län"
  | "Älvsborgs län"
  | "Skaraborgs län"
  | "Värmlands län"
  | "Örebro län"
  | "Västmanlands län"
  | "Kopparbergs län"
  | "Gävleborgs län"
  | "Västernorrlands län"
  | "Jämtlands län"
  | "Västerbottens län"
  | "Norrbottens län"
  | "Unknown";

export function getPlaceOfBirth(digits: string): PlaceOfBirth {
  if (digits.length < 2) throw new Error("Digits must be at least 2 characters long");

  const placeOfBirthDigits = parseInt(digits.substring(0, 2));

  if (placeOfBirthDigits >= 0 && placeOfBirthDigits <= 9) return "Stockholms stad";
  else if (placeOfBirthDigits >= 10 && placeOfBirthDigits <= 13) return "Stockholms län";
  else if (placeOfBirthDigits >= 14 && placeOfBirthDigits <= 15) return "Uppsala län";
  else if (placeOfBirthDigits >= 16 && placeOfBirthDigits <= 18) return "Södermanlands län";
  else if (placeOfBirthDigits >= 19 && placeOfBirthDigits <= 23) return "Östergötlands län";
  else if (placeOfBirthDigits >= 24 && placeOfBirthDigits <= 26) return "Jönköpings län";
  else if (placeOfBirthDigits >= 27 && placeOfBirthDigits <= 28) return "Kronobergs län";
  else if (placeOfBirthDigits >= 29 && placeOfBirthDigits <= 31) return "Kalmar län";
  else if (placeOfBirthDigits === 32) return "Gotlands län";
  else if (placeOfBirthDigits >= 33 && placeOfBirthDigits <= 34) return "Blekinge län";
  else if (placeOfBirthDigits >= 35 && placeOfBirthDigits <= 38) return "Kristianstads län";
  else if (placeOfBirthDigits >= 39 && placeOfBirthDigits <= 45) return "Malmöhus län";
  else if (placeOfBirthDigits >= 46 && placeOfBirthDigits <= 47) return "Hallands län";
  else if (placeOfBirthDigits >= 48 && placeOfBirthDigits <= 54) return "Göteborgs och Bohus län";
  else if (placeOfBirthDigits >= 55 && placeOfBirthDigits <= 58) return "Älvsborgs län";
  else if (placeOfBirthDigits >= 59 && placeOfBirthDigits <= 61) return "Skaraborgs län";
  else if (placeOfBirthDigits >= 62 && placeOfBirthDigits <= 64) return "Värmlands län";
  else if (placeOfBirthDigits >= 66 && placeOfBirthDigits <= 68) return "Örebro län";
  else if (placeOfBirthDigits >= 69 && placeOfBirthDigits <= 70) return "Västmanlands län";
  else if (placeOfBirthDigits >= 71 && placeOfBirthDigits <= 73) return "Kopparbergs län";
  else if (placeOfBirthDigits >= 75 && placeOfBirthDigits <= 77) return "Gävleborgs län";
  else if (placeOfBirthDigits >= 78 && placeOfBirthDigits <= 81) return "Västernorrlands län";
  else if (placeOfBirthDigits >= 82 && placeOfBirthDigits <= 84) return "Jämtlands län";
  else if (placeOfBirthDigits >= 85 && placeOfBirthDigits <= 88) return "Västerbottens län";
  else if (placeOfBirthDigits >= 89 && placeOfBirthDigits <= 92) return "Norrbottens län";
  else return "Unknown";
}
